mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+
+        param_sig_keys = [
+            key for key in param_sig_keys
+            if param_sig[key].kind in [
+                inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD
+            ]
+        ]
+
+        if self_obj is not None:
+            call_fn_name = f'{self_obj.__class__.__name__}.{metric_fn.__name__}'
+        else:
+            call_fn_name = metric_fn.__name__
+
+        assert len(param_sig_keys) >= len(call_pos_args), 'Found too many positional arguments.'
+        for arg_name, arg in zip(param_sig_keys, call_pos_args):
+            arg_list.append(f'{arg_name}={arg_to_str(arg)}')
+
+        for arg_name, arg in call_kwargs.items():
+            arg_list.append(f'{arg_name}={arg_to_str(arg)}')
+
+        arg_list_str = ', '.join(arg_list)
+
+        return f'{call_fn_name}({arg_list_str})'
+
+    def register_metric_info(self, run_id, metric_name, dataset_name, call_command):
+
+        call_cmd_list = self._metric_api_call_info[run_id][metric_name]
+
+        index = len(call_cmd_list)
+        metric_name_with_index = self.gen_name_with_index(metric_name, index)
+        metric_key = f'{metric_name_with_index}_{dataset_name}'
+
+        call_cmd_list.append((metric_key, call_command))
+
+        self._metric_info_artifact_need_update[run_id] = True
+        return metric_key
+
+    def register_metric_api_call(self, metric_name, call_command, call_pos_args, call_kwargs):
+        """
+        Given a metric api call (include the called metric function, and call arguments)
+        Register the call information (arguments dict) into the `metric_api_call_arg_dict_list_map`
+        and return a tuple of (is_ok, run_id, metric_key)
+        - metric key format is: {eval_dataset_name}_{metric_name}
+          metric_name is generated by metric function name, if there're multiple calls with
+          different call arguments, then appending index to the name.
+        """
+        call_arg_list = list(call_pos_args) + list(call_kwargs.values())
+
+        dataset_id_list = self._pred_result_id_to_dataset_name_and_run_id.keys()
+
+        # Note: some metric API the arguments is not like `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        for arg in call_arg_list:
+            if arg is not None and not np.isscalar(arg) and id(arg) in dataset_id_list:
+                dataset_name, run_id = self._pred_result_id_to_dataset_name_and_run_id[id(arg)]
+                break
+        else:
+            return False, None, None
+
+        metric_key = self.register_metric_info(
+            run_id, metric_name, dataset_name, call_command
+        )
+
+        return True, run_id, metric_key
+
+    def log_eval_metric(self, run_id, key, value):
```

Comment:
This is named pretty similarly to `eval_and_log_metrics`. Can we call this `log_post_training_metric`?

Reply:
Can we also add a docstring for this method?